components/powersensor/renogy/renogy.go

```diff
@@ -0,0 +1,300 @@
+
+	// add all readings.
+	r.addReading(client, solarVoltReg, 1, "SolarVolt")
+	r.addReading(client, solarAmpReg, 2, "SolarAmp")
+	r.addReading(client, solarWattReg, 0, "SolarWatt")
+	r.addReading(client, loadVoltReg, 1, "LoadVolt")
+	r.addReading(client, loadAmpReg, 2, "LoadAmp")
+	r.addReading(client, loadWattReg, 0, "LoadWatt")
+	r.addReading(client, battVoltReg, 1, "BattVolt")
+	r.addReading(client, battChargePctReg, 0, "BattChargePct")
+	r.addReading(client, maxSolarTodayWattReg, 0, "MaxSolarTodayWatt")
+	r.addReading(client, minSolarTodayWattReg, 0, "MinSolarTodayWatt")
+	r.addReading(client, maxBattTodayVoltReg, 1, "MaxBattTodayVolt")
+	r.addReading(client, minBattTodayVoltReg, 1, "MinBattTodayVolt")
+	r.addReading(client, maxSolarTodayAmpReg, 2, "MaxSolarTodayAmp")
+	r.addReading(client, minSolarTodayAmpReg, 1, "MinSolarTodayAmp")
+	r.addReading(client, chargeTodayAmpHrsReg, 0, "ChargeTodayAmpHrs")
+	r.addReading(client, dischargeTodayAmpHrsReg, 0, "DischargeTodayAmpHrs")
+	r.addReading(client, chargeTodayWattHrsReg, 0, "ChargeTodayWattHrs")
+	r.addReading(client, dischargeTodayWattHrsReg, 0, "DischargeTodayWattHrs")
+	r.addReading(client, totalBattOverChargesReg, 0, "TotalBattOverCharges")
+	r.addReading(client, totalBattFullChargesReg, 0, "TotalBattFullCharges")
+
+	// Controller and battery temperates require math on controller deg register.
+	tempReading, err := r.readRegister(client, controllerDegCReg, 0)
+	if err != nil {
+		return readings, err
+	}
+
+	battTempSign := (int16(tempReading) & 0b0000000010000000) >> 7
+	battTemp := int16(tempReading) & 0b0000000001111111
+	if battTempSign == 1 {
+		battTemp = -battTemp
+	}
+
+	readings["BattDegC"] = int32(battTemp)
+
+	ctlTempSign := (int32(tempReading) & 0b1000000000000000) >> 15
+	ctlTemp := (int16(tempReading) & 0b0111111100000000) >> 8
+	if ctlTempSign == 1 {
+		ctlTemp = -ctlTemp
+	}
+	readings["ControllerDegC"] = int32(ctlTemp)
+
+	err = r.handler.Close()
+	if err != nil {
+		return readings, err
+	}
+	r.handler = nil
+	return readings, nil
+}
+
+func (r *Renogy) addReading(client modbus.Client, register uint16, precision uint, reading string) {
+	value, err := r.readRegister(client, register, precision)
+	if err != nil {
+		r.logger.Errorf("error getting reading: %s : %v", reading, err)
+	} else {
+		readings[reading] = value
+	}
+}
+
+func (r *Renogy) readRegister(client modbus.Client, register uint16, precision uint) (result float32, err error) {
+	r.mu.Lock()
+	b, err := client.ReadHoldingRegisters(register, 1)
+	r.mu.Unlock()
+	if err != nil {
+		return 0, err
+	}
+	if len(b) > 0 {
+		result = float32FromBytes(b, precision)
+	} else {
+		result = 0
+	}
+	return result, nil
+}
+
+func float32FromBytes(bytes []byte, precision uint) float32 {
+	i := binary.BigEndian.Uint16(bytes)
+	ratio := math.Pow(10, float64(precision))
+	return float32(float64(i) / ratio)
+}
+
+// Close closes the renogy modbus
+func (r *Renogy) Close(ctx context.Context) error {
+	r.mu.Lock()
```

Comment:
preferred pattern is 
```go
mu.Lock()
defer mu.Unlock()
```

to return at teh function return in go, unless we need to do some special things. like with the clients in readRegister.